src/oci/runtime/runc.rs

```diff
@@ -0,0 +1,84 @@
+//! runc OCI runtime implementation
+
+use crate::oci::runtime::{Container, ContainerState, ContainerStats, Runtime};
+use crate::oci::spec::runtime::{LinuxResources, Spec};
+use anyhow::Result;
+use async_trait::async_trait;
+use derive_builder::Builder;
+use getset::Getters;
+use tokio::{process::Command, signal::unix::SignalKind};
+
+struct Runc;
+
+#[async_trait]
+impl Runtime for Runc {
+    type Container = RuncContainer;
+
+    /// Create a new container.
+    async fn create(&mut self) -> Result<RuncContainer> {
+        unimplemented!()
+    }
+
+    /// Execute the user defined process in a created container.
+    async fn start(&mut self, _container: &mut RuncContainer) -> Result<()> {
+        unimplemented!()
+    }
+
+    /// Delete any resources held by the container often used with detached container.
+    async fn delete(&mut self, _container: &mut RuncContainer) -> Result<()> {
+        unimplemented!()
+    }
+
+    /// Suspend all processes inside the container.
+    async fn pause(&mut self, _container: &mut RuncContainer) -> Result<()> {
+        unimplemented!()
+    }
+
+    /// Resumes all processes that have been previously paused.
+    async fn resume(&mut self, _container: &mut RuncContainer) -> Result<()> {
+        unimplemented!()
+    }
+
+    /// Send the specified signal to the container's init process.
+    async fn kill(
+        &mut self,
+        _container: &mut RuncContainer,
+        _signal_kind: SignalKind,
+    ) -> Result<()> {
+        unimplemented!()
+    }
+
+    /// Update container resource constraints.
+    async fn update(
+        &mut self,
+        _container: &mut RuncContainer,
+        _resources: &LinuxResources,
+    ) -> Result<()> {
+        unimplemented!()
+    }
+
+    /// Execute the provided process inside the container.
+    async fn exec(&self, _container: &RuncContainer, _command: &Command) -> Result<()> {
+        unimplemented!()
+    }
+
+    /// Retrieve container resource statistics.
+    async fn stats(&self, _container: &RuncContainer) -> Result<ContainerStats> {
+        unimplemented!()
+    }
+
+    /// Retrieve the state of a container.
+    async fn state(&self, _container: &RuncContainer) -> Result<ContainerState> {
+        unimplemented!()
+    }
+}
+
+#[derive(Debug, Default, Builder, Getters)]
+#[builder(default, pattern = "owned", setter(into, strip_option))]
+/// A runc container abstraction.
+pub struct RuncContainer {
```

Comment:
We can call this `OCIContainer` or similar and support `runc` and `crun` together :)

Reply:
Sounds good! Do you think we even need the associated type on the trait? We could also make `Container` a `struct` (not a `trait`) and pre-define the data we need, like the `spec`. Not sure if this decreases the flexibility.

I also renamed the `Runc` implementation to `Cli`. That's more generic and we can tell the user that this implementation is for CLI-compatible OCI runtimes, like runc and crun.

Reply:
I think we can use the associate type :+1: and maybe aggregate the concrete ones with a `BaseContainer` type that has information common to all container types.

Reply:
With the renaming of `Runtime` to `Container` we can directly refer to `Self` as type. Now I do not see any reason for the associated type any more.

The `create` method is gone in replacement of the builder pattern, too.